src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.Async.Stream.cs

```diff
@@ -15,42 +15,50 @@ public static partial class JsonConverter
     {
         private const int HalfMaxValue = int.MaxValue / 2;
 
-        public static Task<T> FromJsonAsync<T>(this Stream source, JsonConverterSettings options = null, CancellationToken cancellationToken = default)
+        public static Task<T> FromJsonAsync<T>(this Stream source, JsonConverterSettings settings = null, CancellationToken cancellationToken = default)
         {
             if (source == null)
                 throw new ArgumentNullException(nameof(source));
 
-            return FromJsonAsync<T>(source, typeof(T), options, cancellationToken);
+            return FromJsonAsync<T>(source, typeof(T), settings, cancellationToken);
         }
 
-        public static Task<object> FromJsonAsync(this Stream source, Type returnType, JsonConverterSettings options = null, CancellationToken cancellationToken = default)
+        public static Task<object> FromJsonAsync(this Stream source, Type returnType, JsonConverterSettings settings = null, CancellationToken cancellationToken = default)
         {
             if (source == null)
                 throw new ArgumentNullException(nameof(source));
 
             if (returnType == null)
                 throw new ArgumentNullException(nameof(returnType));
 
-            return FromJsonAsync<object>(source, returnType, options, cancellationToken);
+            return FromJsonAsync<object>(source, returnType, settings, cancellationToken);
         }
 
-        private static async Task<T> FromJsonAsync<T>(this Stream source, Type returnType, JsonConverterSettings options = null, CancellationToken cancellationToken = default)
+        private static async Task<T> FromJsonAsync<T>(this Stream source, Type returnType, JsonConverterSettings settings = null, CancellationToken cancellationToken = default)
         {
             if (source == null)
                 throw new ArgumentNullException(nameof(source));
 
-            if (options == null)
-                options = s_DefaultSettings;
+            if (settings == null)
+                settings = s_DefaultSettings;
 
-            var readerState = new JsonReaderState(options.MaxDepth, options.ReaderOptions);
-            JsonObjectState current = default;
-            List<JsonObjectState> previous = null;
+            FromJsonObjectState current = default;
+            JsonClassInfo classInfo = settings.GetOrAddClass(returnType);
+            current.ClassInfo = classInfo;
+            if (classInfo.ClassType != ClassType.Object)
+            {
+                current.PropertyInfo = classInfo.GetPolicyProperty();
+            }
+
+            var readerState = new JsonReaderState(settings.MaxDepth, settings.ReaderOptions);
+            List<FromJsonObjectState> previous = null;
             int arrayIndex = 0;
 
             int bytesRemaining = 0;
             int bytesRead;
 
-            byte[] buffer = ArrayPool<byte>.Shared.Rent(options.DefaultBufferSize);
+            // todo: switch to IBufferWriter implementation.
```

Comment:
Which implementation?

Reply:
The prototype implementation of IBufferWriter that you added - ArrayBufferWriter. If it helps remove the allocation and pool logic here.